HARK/core.py

```diff
@@ -27,11 +28,256 @@
+
+        Parameters
+        ----------
+        key : str
+            Name of parameter
+        value : Any
+            Value of parameter
+        """
+        if key.startswith("_"):
+            # Handle setting internal attributes normally
+            super().__setattr__(key, value)
+        else:
+            # Handle setting parameters as dictionary items
+            new_value = self.__infer_dims__(key, value)
+            super().__setattr__(key, new_value)
+
+    def keys(self):
+        """
+        Returns a list of the names of the parameters.
+        """
+        return self._age_inv + self._age_var
+
+    def values(self):
+        """
+        Returns a list of the values of the parameters.
+        """
+        return [getattr(self, key) for key in self.keys()]
+
+    def items(self):
+        """
+        Returns a list of tuples of the form (name, value) for each parameter.
+        """
+        return [(key, getattr(self, key)) for key in self.keys()]
+
+    def __iter__(self):
+        """
+        Allows for iterating over the parameter names.
+        """
+        return iter(self.keys())
+
+    def __deepcopy__(self, memo):
+        """
+        Returns a deep copy of the Parameters object.
+        """
+        return Parameters(**deepcopy(self.to_dict(), memo))
+
+    def to_dict(self):
+        """
+        Returns a dictionary of the parameters.
+        """
+        return {key: getattr(self, key) for key in self.keys()}
+
+    def to_namedtuple(self):
+        """
+        Returns a namedtuple of the parameters.
+        """
+        return namedtuple("Parameters", self.keys())(**self.to_dict())
+
+    def update(self, other_params):
+        """
+        Updates the parameters with the values from another
+        Parameters object or a dictionary.
+
+        Parameters
+        ----------
+        other_params : Parameters or dict
+            Parameters object or dictionary of parameters to update with.
+        """
+        if isinstance(other_params, Parameters):
+            for key, value in other_params:
+                setattr(self, key, value)
+        elif isinstance(other_params, dict):
+            for key, value in other_params.items():
+                setattr(self, key, value)
+        else:
+            raise ValueError("Parameters must be a dict or a Parameters object")
+
+    def __str__(self):
+        """
+        Returns a simple string representation of the Parameters object.
+        """
+        return f"Parameters({str(self.to_dict())})"
+
+    def __repr__(self):
+        """
+        Returns a detailed string representation of the Parameters object.
+        """
+        return f"Parameters( _age_inv = {str(self._age_inv)}, _age_var = {str(self._age_var)}, | {str(self.to_dict())})"
+
+
 class Model:
     """
     A class with special handling of parameters assignment.
     """
 
+    def __init__(self, **kwds):
+        self.parameters = Parameters(**kwds)
```

Comment:
I'm thinking of doing some work on the Model class to give it more of the features of the Dolo models, especially holding 'equations':
https://github.com/EconForge/dolo.py/blob/master/examples/models/consumption_savings_iid.yaml

I think that 'Parameter's maps to what Dolo calls `calibration`.

I've noticed that our Model constructor took no arguments before.
Here, the arguments are all being coopted in Parameters. I think that's strictly speaking better.

But what if there's even more to models besides parameter storage?
What should the initializer look like then?

Reply:
While working on this and DistributionOfWealthMPC, I was thinking that we should have Parameters and Options. Parameters includes calibrated specifications of ints, floats, vectors, that are required to solve a particular period/stage of a model, other than functions and input/output states. Options are strings, booleans, and other non-primitive python objects, that might help specify how we solve/simulate a particular stage/period, but are not necessarily mathematical objects required to solve a model. Then, an initializer would look like:

`Model( _abstract_model_structure_, params = Parameters, opts = Options)`